File: NodeRegion.m
%% NodeRegion Class Definition
% A NodeRegion specifies a spatial area over which nodes should be
% generated. The NodeRegion class was created to be able to specify nodes without
% relying on cellular definitions.
%
% 7-May 2011
% Paul Grogan, ptgrogan@mit.edu
%%
classdef NodeRegion < AbstractRegion
    properties(Constant)
        UNDEFINED = 0;      % type not defined
        POLYGON = 1;        % generates nodes within a polygonal region
        POLYLINE = 2;       % generates nodes along a polyline
        POLYPOINT = 3;      % generates nodes at polypoints
    end
    properties
        id;                 % unique identifier for node region
        nodeTypeId;         % node type id for assignment
        layerId;            % layer id for assignment
        type;               % type of node generation in region
        description;        % description of region, string
    end
    methods
        %% NodeRegion Constructor
        % Instantiates a new NodeRegion with specified parameters.
        %
        % obj = NodeRegion(id, nodeTypeId, layerIds, verticesX, verticesY, 
        %           type, description)
        %   id:         unique identifier for node region
        %   nodeTypeId: node type id for assignment
        %   layerIds:   array of layer ids for assignment
        %   verticesX:  array of x-coordinate vertices (counter-clockwise)
        %   verticesY:  array of y-coordinate vertices (counter-clockwise)
        %   type:       constant parameter to define type of generation
        %   description:description of the region, string
        %
        % obj = NodeRegion(nodeTypeId, layerIds, verticesX, verticesY, 
        %           type, description)
        %   nodeTypeId: node type id for assignment
        %   layerIds:   array of layer ids for assignment
        %   verticesX:  array of x-coordinate vertices (counter-clockwise)
        %   verticesY:  array of y-coordinate vertices (counter-clockwise)
        %   type:       constant parameter to define type of generation
        %   description:description of the region, string
        %
        % obj = NodeRegion()
        function obj = NodeRegion(varargin)
            if nargin == 7
                obj.id = varargin{1};
                obj.nodeTypeId = varargin{2};
                obj.layerId = varargin{3};
                obj.verticesX = varargin{4};
                obj.verticesY = varargin{5};
                obj.type = varargin{6};
                obj.description = varargin{7};
            elseif nargin == 6
                obj.id = SynthesisTemplate.instance().GetNextNodeRegionId();
                obj.nodeTypeId = varargin{1};
                obj.layerId = varargin{2};
                obj.verticesX = varargin{3};
                obj.verticesY = varargin{4};
                obj.type = varargin{5};
                obj.description = varargin{6};
            else
                obj.id = SynthesisTemplate.instance().GetNextNodeRegionId();
                obj.nodeTypeId = 0;
                obj.layerId = 0;
                obj.verticesX = 0;
                obj.verticesY = 0;
                obj.type = NodeRegion.UNDEFINED;
                obj.description = '';
            end
        end
        
        %% GenerateNodes Function
        % Generates the nodes within the node region and automatically adds
        % to system definition.
        %
        % GenerateNodes(system)
        %   system: the system within which to generate nodes
        function GenerateNodes(obj,system)
            synthTemp = SynthesisTemplate.instance();
            if obj.type==NodeRegion.POLYGON
                for i=1:length(synthTemp.city.cells)
                    cell = synthTemp.city.cells(i);
                    if obj.ContainsCell(cell)
                        obj.CreateNode(system,cell);
                    end
                end
            elseif obj.type==NodeRegion.POLYLINE
                for i=1:length(synthTemp.city.cells)
                    cell = synthTemp.city.cells(i);
                    for p=1:length(obj.verticesX)-1
                        x1 = obj.verticesX(p);
                        y1 = obj.verticesY(p);
                        x2 = obj.verticesX(p+1);
                        y2 = obj.verticesY(p+1);
                        if cell.IntersectsLine(x1,y1,x2,y2)
                            obj.CreateNode(system,cell);
                            break;
                        end
                    end
                end
            elseif obj.type==NodeRegion.POLYPOINT
                for i=1:length(synthTemp.city.cells)
                    cell = synthTemp.city.cells(i);
                    for p=1:length(obj.verticesX)
                        x = obj.verticesX(p);
                        y = obj.verticesY(p);
                        if cell.ContainsPoint(x,y);
                            obj.CreateNode(system,cell);
                            break;
                        end
                    end
                end
            end
        end
    end
    methods(Access=private)
        %% CreateNode Function
        % Creates a new node at a cell within a system.
        function CreateNode(obj,system,cell)
            synthTemp = SynthesisTemplate.instance();
            system.nodes(end+1) = Node(cell, ...
                synthTemp.city.layers([synthTemp.city.layers.id]==obj.layerId), ...
                synthTemp.nodeTypes([synthTemp.nodeTypes.id]==obj.nodeTypeId));
        end
    end
end

File: src/main/java/edu/mit/citynet/core/NodeRegion.java
/*
 * CityNet: Integrated Urban Development Decision Support Tool
 * 
 * Copyright (c) 2011 MIT Strategic Engineering Research Group
 */
package edu.mit.citynet.core;

import javax.swing.ImageIcon;

import com.vividsolutions.jts.geom.Coordinate;
import com.vividsolutions.jts.geom.GeometryFactory;
import com.vividsolutions.jts.geom.LineString;

import edu.mit.citynet.CityNet;
import edu.mit.citynet.util.CityNetIcon;

/**
 * The NodeRegion class describes nodes that will be generated at cells over a 
 * region of space. It was created to be able to specify nodes without relying
 * on cellular definitions.
 * 
 * @author Paul Grogan, ptgrogan@mit.edu
 */
public class NodeRegion extends AbstractRegion implements Cloneable {
	
	/**
	 * The NodeRegionType enumeration. Defines the set of possible regions that
	 * can be specified, each with differing node generation behaviors.
	 */
	public enum NodeRegionType {
		UNDEFINED("Undefined"),
		POLYGON("Polygon"),
		POLYLINE("Polyline"),
		POLYPOINT("Polypoint");

		private String name;
		
		/**
		 * Instantiates a new node region type.
		 *
		 * @param name the name
		 */
		private NodeRegionType(String name) {
			this.name = name;
		}
		
		/**
		 * Gets the name.
		 *
		 * @return the name
		 */
		public String getName() {
			return name;
		}
		
		/* (non-Javadoc)
		 * @see java.lang.Enum#toString()
		 */
		public String toString() {
			return getName();
		}
		
		/**
		 * Gets the icon.
		 *
		 * @return the icon
		 */
		public ImageIcon getIcon() {
			switch(this) {
			case POLYGON: return CityNetIcon.POLYGON_REGION.getIcon();
			case POLYLINE: return CityNetIcon.POLYLINE_REGION.getIcon();
			case POLYPOINT: return CityNetIcon.POLYPOINT_REGION.getIcon();
			default: return CityNetIcon.UNDEFINED_REGION.getIcon();
			}
		}
		
		/**
		 * Gets the node region type.
		 *
		 * @param name the name
		 * @return the node region type
		 */
		public static NodeRegionType getNodeRegionType(String name) {
			for(NodeRegionType t : NodeRegionType.values()) {
				if(t.getName().toLowerCase().equals(name.toLowerCase())) return t;
			}
			return UNDEFINED;
		}
	}
	private int id;
	private NodeRegionType nodeRegionType;
	private Layer layer;
	private NodeType nodeType;
	
	/**
	 * Instantiates a new node region.
	 */
	public NodeRegion() {
		super();
		setDescription("New Node Region");
	}
	
	/**
	 * Generate nodes.
	 *
	 * @param system the system
	 * @return the set of generated nodes
	 */
	public void generateNodes(CitySystem system) {
		GeometryFactory gf = CityNet.getInstance().getGeometryFactory();
		switch(nodeRegionType) {
		case POLYGON:
			for(Cell cell : CityNet.getInstance().getCity().getCells()) {
				if(containsPolygon(cell.getPolygon(), 0.5)) {
					createNode(system,cell);
				}
			}
			break;
		case POLYLINE:
			LineString line = gf.createLineString(getCoordinateList().toCoordinateArray());
			for(Cell cell : CityNet.getInstance().getCity().getCells()) {
				if(cell.intersectsLine(line)) {
					createNode(system,cell);
				}
			}
			break;
		case POLYPOINT:
			for(Cell cell : CityNet.getInstance().getCity().getCells()) {
				for(Coordinate coord : getCoordinateList().toCoordinateArray()) {
					if(cell.containsPoint(gf.createPoint(coord))) {
						createNode(system,cell);
						break;
					}
				}
			}
			break;
		case UNDEFINED:
			throw new IllegalStateException("Unknown or undefined node region type.");
		}
	}
	
	private void createNode(CitySystem system, Cell cell) {
		for(Node n : system.getNodes()) {
			if(n.getCell().equals(cell) && n.getLayer().equals(layer)) {
				return; // node exists at same location
			}
		}
		Node node = new Node();
		node.setId(CityNet.getInstance().getNextNodeId());
		node.setCell(cell);
		node.setLayer(layer);
		node.setNodeType(nodeType);
		system.addNode(node);
	}
	
	/**
	 * Gets the id.
	 *
	 * @return the id
	 */
	public int getId() {
		return id;
	}

	/**
	 * Sets the id.
	 *
	 * @param id the new id
	 */
	public void setId(int id) {
		this.id = id;
	}

	/**
	 * Gets the node region type.
	 *
	 * @return the node region type
	 */
	public NodeRegionType getNodeRegionType() {
		return nodeRegionType;
	}

	/**
	 * Sets the node region type.
	 *
	 * @param nodeRegiontype the new node region type
	 */
	public void setNodeRegionType(NodeRegionType nodeRegionType) {
		this.nodeRegionType = nodeRegionType;
	}

	/**
	 * Gets the layer.
	 *
	 * @return the layer
	 */
	public Layer getLayer() {
		return layer;
	}

	/**
	 * Sets the layer.
	 *
	 * @param layer the new layer
	 */
	public void setLayer(Layer layer) {
		this.layer = layer;
	}

	/**
	 * Gets the node type.
	 *
	 * @return the node type
	 */
	public NodeType getNodeType() {
		return nodeType;
	}

	/**
	 * Sets the node type.
	 *
	 * @param nodeType the new node type
	 */
	public void setNodeType(NodeType nodeType) {
		this.nodeType = nodeType;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	public boolean equals(Object object) {
		if(object instanceof NodeRegion) {
			return id==((NodeRegion)object).getId();
		} else return false;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	public String toString() {
		return "NodeRegion" + id;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#clone()
	 */
	public NodeRegion clone() {
		NodeRegion clone = new NodeRegion();
		clone.setId(0);
		clone.setDescription(getDescription() + " (Copy)");
		clone.setLayer(getLayer());
		clone.setNodeRegionType(getNodeRegionType());
		clone.setNodeType(getNodeType());
		clone.setCoordinateList(getCoordinateList());
		return clone;
	}
}